block/src/Pos/Block/Logic/VAR.hs

```diff
@@ -179,12 +185,34 @@ verifyAndApplyBlocks rollback blocks = runExceptT $ do
                 logDebug "Rolling: Verification done, applying unsafe block"
                 lift $ applyBlocksUnsafe (ShouldCallBListener True) newBlunds (Just pModifier)
                 case getOldestFirst suffix of
-                    [] -> GS.getTip
+                    [] -> do
+                        logDebug "Rolling: Applying done"
+                        GS.getTip
                     (genesis:xs) -> do
                         logDebug "Rolling: Applying done, next portion"
                         rollingVerifyAndApply (toNewestFirst newBlunds : blunds) $
                             spanEpoch (OldestFirst (genesis:|xs))
 
+-- | 'verifyAndApplyBlocks' in form of a conduit.
+--
+-- We simply break blocks into chunks of 1000 and then do
+-- 'verifyAndApplyBlocks' for each. This isn't the most optimal strategy but
+-- it'll do for now.
+--
+-- If there are no blocks, 'verifyAndApplyBlocksC' returns the current tip.
+verifyAndApplyBlocksC
+    :: forall ctx m. (BlockLrcMode ctx m, MonadMempoolNormalization ctx m)
+    => Bool -> Consumer Block m (Either ApplyBlocksException HeaderHash)
+verifyAndApplyBlocksC rollback = do
+    tip <- lift GS.getTip
+    splitC 1000 .| Conduit.mapC OldestFirst .| processBatches tip
+  where
+    processBatches tip = Conduit.await >>= \case
+        Nothing     -> pure (Right tip)
+        Just blocks -> lift (verifyAndApplyBlocks rollback blocks) >>= \case
+            Left err     -> pure (Left err)
```

Comment:
What state does `verifyAndApplyBlocks` leave if it errors?

Reply:
If `rollback` is true, it does a rollback. If `rollback` is false, it silently doesn't apply failed blocks and returns the current tip.

Reply:
So if `rollback` is `False` and there is an error, we are left with a state where `1000*k` blocks have been applied, where `k` is some integer?

Reply:
No, it can be any amount of blocks.